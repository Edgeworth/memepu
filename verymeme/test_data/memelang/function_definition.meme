enum Unit {}

fn bubblesort<T>(v: const* vector<*T>) Unit {
  for (var i: i32 = 0; i < v.size(); i++) {
    for (var j: i32 = 0; j < v.size() - 1; j++) {
      if (v[j] > v[j + 1]) {
        swap(v[j], v[j + 1]);
      }
    }
  }
}

fn main() i32 {
  var v: auto = vector<int8>.new();
  v.push(4);
  v.push(3);
  v.push(1);
  bubblesort(&v);
  return 0;
}
