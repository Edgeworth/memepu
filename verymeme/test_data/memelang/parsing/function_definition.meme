enum Unit {}

fn bubblesort<T>(v: const* vector<*T>) Unit {
  for (var i: i32 = 0; i < v.size(); i++) {
    for (var j: i32 = 0; j < v.size() - 1; j++) {
      if (v[j] > v[j + 1]) {
        swap(v[j], v[j + 1]);
      }
    }
  }
}
