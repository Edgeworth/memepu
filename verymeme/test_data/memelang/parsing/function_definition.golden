File
 FnDefn
  FnSig
   Typename(bubblesort)
    Typelist(T)
   VarDecl
    VarRef(v)
    Type(vector)
     Qualifier(const: 1 ptr: 1, array: 0)
     Type(T)
      Qualifier(const: 0 ptr: 1, array: 0)
   Type(Unit)
  StmtBlk
   For
    Var
     VarDecl
      VarRef(i)
      Type(i32)
     IntLit(0)
    Op(LT)
     VarRef(i)
     Op(FN_CALL)
      Op(MEMBER_ACCESS)
       VarRef(v)
       VarRef(size)
      FnCallArgs
    Op(POSTFIX_INC)
     VarRef(i)
    StmtBlk
     For
      Var
       VarDecl
        VarRef(j)
        Type(i32)
       IntLit(0)
      Op(LT)
       VarRef(j)
       Op(SUB)
        Op(FN_CALL)
         Op(MEMBER_ACCESS)
          VarRef(v)
          VarRef(size)
         FnCallArgs
        IntLit(1)
      Op(POSTFIX_INC)
       VarRef(j)
      StmtBlk
       If
        Op(GT)
         Op(ARRAY_ACCESS)
          VarRef(v)
          VarRef(j)
         Op(ARRAY_ACCESS)
          VarRef(v)
          Op(ADD)
           VarRef(j)
           IntLit(1)
        StmtBlk
         Op(FN_CALL)
          VarRef(swap)
          FnCallArgs
           Op(ARRAY_ACCESS)
            VarRef(v)
            VarRef(j)
           Op(ARRAY_ACCESS)
            VarRef(v)
            Op(ADD)
             VarRef(j)
             IntLit(1)
 Enum(0 untyped)
  Typename(Unit)
